File: app/src/main/java/co/edu/unipiloto/myapplication/db/SolicitudRepository.java
package co.edu.unipiloto.myapplication.db;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.ArrayList;
import java.util.List;

import co.edu.unipiloto.myapplication.model.Solicitud;

public class SolicitudRepository {
    private final DBHelper helper;
    public static final String EST_PENDIENTE = "PENDIENTE";
    public static final String EST_ASIGNADA  = "ASIGNADA";
    public static final String EST_EN_CAMINO = "EN_CAMINO";
    public static final String EST_ENTREGADA = "ENTREGADA";
    public static final String EST_CONFIRMADA= "CONFIRMADA";
    public static final String EST_CANCELADA = "CANCELADA";

    public SolicitudRepository(Context ctx){ this.helper = new DBHelper(ctx); }

    /**
     * Crear solicitud: ahora guarda también una fila mínima en "direcciones" y referencia su id.
     * Esto mantiene la columna textual "direccion" para compatibilidad.
     */
    public long crear(long userId, String dir, String fecha, String franja, String notas, String zona){
        SQLiteDatabase db = helper.getWritableDatabase();

        // 1) crear entrada en direcciones (guardamos full_address para fallback)
        long direccionId = -1L;
        try {
            ContentValues dirCv = new ContentValues();
            dirCv.put("full_address", dir != null ? dir : "");
            dirCv.put("created_at", System.currentTimeMillis());
            direccionId = db.insert("direcciones", null, dirCv);
        } catch (Exception ex) {
            // si falla la inserción en direcciones, continuamos sin dirección_id (no bloqueante)
            direccionId = -1L;
        }

        // 2) crear solicitud usando la columna textual y la referencia direccion_id (si disponible)
        ContentValues cv = new ContentValues();
        cv.put("user_id", userId);
        cv.put("direccion", dir != null ? dir : "");
        if (direccionId != -1L) cv.put("direccion_id", direccionId);
        cv.put("fecha", fecha);
        cv.put("franja", franja);
        cv.put("notas", notas);
        cv.put("zona", zona);
        cv.put("estado", EST_PENDIENTE);
        cv.put("created_at", System.currentTimeMillis());
        // guia_id queda null al crear
        long newRowId = db.insert("solicitudes", null, cv);
        db.close();
        return newRowId;
    }

    /**
     * Listar solicitudes del cliente (para mostrar en su panel).
     */
    public List<SolicitudItem> listarPorUsuario(long userId){
        SQLiteDatabase db = helper.getReadableDatabase();
        Cursor c = db.rawQuery(
                "SELECT id, direccion, fecha, franja, estado, created_at, zona FROM solicitudes " +
                        "WHERE user_id=? ORDER BY created_at DESC", new String[]{String.valueOf(userId)});
        List<SolicitudItem> list = new ArrayList<>();
        try {
            while(c.moveToNext()){
                SolicitudItem it = new SolicitudItem();
                it.id = c.getLong(0);
                it.direccion = c.getString(1);
                it.fecha = c.getString(2);
                it.franja = c.getString(3);
                it.estado = c.getString(4);
                it.createdAt = c.getLong(5);
                it.zona = c.getString(6);
                list.add(it);
            }
        } finally {
            c.close();
            db.close();
        }
        return list;
    }

    /**
     * Obtener solicitudes (modelo completo) por zona: solo PENDIENTES para que el recolector las vea.
     */
    public List<Solicitud> pendientesPorZona(String zona) {
        SQLiteDatabase db = helper.getReadableDatabase();
        List<Solicitud> lista = new ArrayList<>();
        Cursor c = db.rawQuery(
                "SELECT id, user_id, direccion, fecha, franja, notas, estado, zona, created_at " +
                        "FROM solicitudes WHERE zona=? AND (estado IS NULL OR estado='PENDIENTE') ORDER BY created_at DESC",
                new String[]{zona}
        );
        try {
            while (c.moveToNext()) {
                Solicitud s = new Solicitud(
                        c.getLong(0),
                        c.getLong(1),
                        c.getString(2),
                        c.getString(3),
                        c.getString(4),
                        c.getString(5),
                        c.getString(6),
                        c.getString(7),
                        c.getLong(8)
                );
                lista.add(s);
            }
        } finally {
            c.close();
            db.close();
        }
        return lista;
    }

    /**
     * Obtener solicitudes asignadas a un recolector (útil para DriverDashboardActivity.loadAssignedRoutes)
     */
    public List<Solicitud> getAssignedRoutesByDriver(long recolectorId) {
        return asignadasA(recolectorId);
    }

    /**
     * Asignar/aceptar una solicitud por parte de un recolector.
     * Solo actualiza si la solicitud todavía está PENDIENTE (evita race conditions).
     */
    public int asignarARecolector(long solicitudId, long recolectorId) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("recolector_id", recolectorId);
        cv.put("estado", EST_ASIGNADA);
        // solo actualizar si estaba pendiente
        int rows = db.update("solicitudes", cv, "id=? AND (estado IS NULL OR estado=?)",
                new String[]{String.valueOf(solicitudId), EST_PENDIENTE});
        db.close();
        return rows;
    }

    /**
     * Cancelar solicitud por parte del cliente. Solo si pertenece al usuario y está pendiente.
     */
    public int cancelarSolicitud(long solicitudId, long userId) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("estado", EST_CANCELADA);
        int rows = db.update("solicitudes", cv, "id=? AND user_id=? AND (estado IS NULL OR estado=?)",
                new String[]{String.valueOf(solicitudId), String.valueOf(userId), EST_PENDIENTE});
        db.close();
        return rows;
    }

    /**
     * listar solicitudes asignadas a un recolector
     */
    public List<Solicitud> asignadasA(long recolectorId) {
        SQLiteDatabase db = helper.getReadableDatabase();
        List<Solicitud> out = new ArrayList<>();
        Cursor c = db.rawQuery(
                "SELECT id, user_id, direccion, fecha, franja, notas, estado, zona, created_at " +
                        "FROM solicitudes WHERE recolector_id=? ORDER BY created_at DESC",
                new String[]{String.valueOf(recolectorId)}
        );
        try {
            while (c.moveToNext()) {
                out.add(new Solicitud(
                        c.getLong(0), c.getLong(1), c.getString(2), c.getString(3),
                        c.getString(4), c.getString(5), c.getString(6), c.getString(7), c.getLong(8)
                ));
            }
        } finally {
            c.close();
            db.close();
        }
        return out;
    }

    // --------------------------------------------------------------------------------------
    // 🆕 FUNCIÓN DE RASTREO: Obtener Solicitud por ID 🆕
    // --------------------------------------------------------------------------------------

    /**
     * Obtiene una Solicitud completa por su ID (usado como código de guía en la búsqueda del cliente).
     * @param solicitudId El ID único de la solicitud.
     * @return El objeto Solicitud o null si no se encuentra.
     */
    public Solicitud getSolicitudById(long solicitudId) {
        SQLiteDatabase db = helper.getReadableDatabase();
        Solicitud s = null;
        Cursor c = db.rawQuery(
                "SELECT id, user_id, direccion, fecha, franja, notas, estado, zona, created_at " +
                        "FROM solicitudes WHERE id=?",
                new String[]{String.valueOf(solicitudId)}
        );

        try {
            if (c.moveToFirst()) {
                s = new Solicitud(
                        c.getLong(0),   // id
                        c.getLong(1),   // user_id
                        c.getString(2), // direccion
                        c.getString(3), // fecha
                        c.getString(4), // franja
                        c.getString(5), // notas
                        c.getString(6), // estado
                        c.getString(7), // zona
                        c.getLong(8)    // created_at
                );
            }
        } finally {
            c.close();
            db.close();
        }
        return s;
    }

    // --------------------------------------------------------------------------------------
    // 🆕 NUEVA FUNCIONALIDAD: Actualizar Estado y Guardar Código de Confirmación 🆕
    // --------------------------------------------------------------------------------------

    /**
     * Actualiza el estado a EN_CAMINO y guarda el código de confirmación de 4 dígitos.
     * @param solicitudId ID de la solicitud.
     * @param status El nuevo estado (debería ser EST_EN_CAMINO).
     * @param code El código de 4 dígitos generado.
     * @return Número de filas afectadas (1 si fue exitoso).
     */
    public int updateStatusAndCode(long solicitudId, String status, String code) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();

        // 1. Datos a actualizar
        cv.put("estado", status); // EST_EN_CAMINO
        cv.put("en_camino_at", System.currentTimeMillis());
        cv.put("confirmation_code", code); // Guarda el código de 4 dígitos

        // Condición: La solicitud debe estar pendiente o asignada para iniciar el trayecto
        int rows = db.update("solicitudes", cv,
                "id=? AND (estado=? OR estado=?)",
                new String[]{
                        String.valueOf(solicitudId),
                        EST_ASIGNADA,
                        EST_PENDIENTE
                });

        db.close();
        return rows;
    }


    public static class SolicitudItem {
        public long id; public String direccion; public String fecha; public String franja;
        public String estado; public long createdAt; public String zona;
    }

    // Método genérico para actualizar estado (mantenido por compatibilidad)
    public int actualizarEstado(long solicitudId, String nuevoEstado, String timestampField) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("estado", nuevoEstado);
        if (timestampField != null) {
            cv.put(timestampField, System.currentTimeMillis());
        }
        int rows = db.update("solicitudes", cv, "id=?", new String[]{ String.valueOf(solicitudId) });
        db.close();
        return rows;
    }

    /** Métodos concretos para conveniencia (mantenidos, pero updateStatusAndCode es preferido para SMS) */
    public int marcarEnCamino(long solicitudId, long recolectorId) {
        // NOTA: Este método NO guarda el confirmation_code. Usa updateStatusAndCode para SMS.
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("estado", EST_EN_CAMINO);
        cv.put("recolector_id", recolectorId);
        cv.put("en_camino_at", System.currentTimeMillis());
        int rows = db.update("solicitudes", cv, "id=? AND (estado IS NULL OR estado=? OR estado=?)",
                new String[]{ String.valueOf(solicitudId), EST_PENDIENTE, EST_ASIGNADA });
        db.close();
        return rows;
    }

    public int marcarEntregada(long solicitudId, long recolectorId) {
        SQLiteDatabase db = helper.getWritableDatabase();
        ContentValues cv = new ContentValues();
        cv.put("estado", EST_ENTREGADA);
        cv.put("entregado_at", System.currentTimeMillis());
        cv.put("recolector_id", recolectorId);
        int rows = db.update("solicitudes", cv,
                "id=? AND (estado=? OR estado=?)",
                new String[]{ String.valueOf(solicitudId), EST_ASIGNADA, EST_EN_CAMINO });
        db.close();
        return rows;
    }
}

File: app/src/main/java/co/edu/unipiloto/myapplication/db/SolicitudRepository.kt
package co.edu.unipiloto.myapplication.db

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.util.Log
// 🏆 IMPORTAR el modelo de datos ENRIQUECIDO
import co.edu.unipiloto.myapplication.models.Solicitud
import co.edu.unipiloto.myapplication.db.SolicitudRepository.Companion.TABLE_SOLICITUDES
import java.util.UUID // Necesario para generar un tracking_number único

/**
 * Repositorio encargado de gestionar las operaciones CRUD (Crear, Leer, Actualizar, Borrar)
 * para las solicitudes de envío, así como sus dependencias (direcciones y guías).
 */
class SolicitudRepository(context: Context) {

    private val helper: DBHelper = DBHelper(context)
    private val TAG = "SolicitudRepo"

    companion object {
        // Constantes para los nombres de las tablas para mejorar la legibilidad del SQL
        private const val TABLE_SOLICITUDES = "solicitudes"
        private const val TABLE_DIRECCIONES = "direcciones"
        private const val TABLE_USERS = "users"
        private const val TABLE_GUIA = "guia"
    }

    // ==========================================================
    // 1. CREACIÓN (Cliente)
    // ==========================================================

    /**
     * Crea una nueva solicitud de recolección, insertando la dirección, la guía (paquete) y la solicitud en una transacción.
     *
     * @return El ID de la nueva solicitud insertada, o -1L si la transacción falla.
     */
    fun crearSolicitud(
        userId: Long,
        direccionCompleta: String,
        ciudad: String,
        latitudRecoleccion: Double,
        longitudRecoleccion: Double,
        peso: Double,
        precio: Double,
        fechaRecoleccion: String,
        franjaHoraria: String,
        notas: String?,
        zona: String
    ): Long {
        val db = helper.writableDatabase
        var newSolicitudId: Long = -1L
        var newDireccionId: Long = -1L
        var newGuiaId: Long = -1L // Nuevo ID de Guía

        db.beginTransaction()
        try {
            // 1. CREAR LA DIRECCIÓN ASOCIADA
            val cvDireccion = ContentValues().apply {
                put("user_id", userId)
                put("ciudad", ciudad)
                put("full_address", direccionCompleta)
                put("created_at", System.currentTimeMillis())
                put("latitude", latitudRecoleccion) // <-- INSERTAR LATITUD
                put("longitude", longitudRecoleccion)
            }
            newDireccionId = db.insert(TABLE_DIRECCIONES, null, cvDireccion)

            if (newDireccionId == -1L) throw Exception("Error al insertar la dirección.")

            // 2. CREAR LA GUÍA (PAQUETE)
            val trackingNumber = UUID.randomUUID().toString().substring(0, 10).uppercase()
            val cvGuia = ContentValues().apply {
                put("tracking_number", trackingNumber)
                put("descripcion", notas) // Usamos las notas como descripción del contenido
                put("valor", precio)
                put("peso", peso)
                put("created_at", System.currentTimeMillis())
            }
            newGuiaId = db.insert(TABLE_GUIA, null, cvGuia)

            if (newGuiaId == -1L) throw Exception("Error al insertar la guía.")


            // 3. CREAR LA SOLICITUD
            val cvSolicitud = ContentValues().apply {
                put("user_id", userId)
                put("direccion_id", newDireccionId)
                put("guia_id", newGuiaId) // <-- ASOCIAR EL ID DE LA GUÍA
                put("fecha", fechaRecoleccion)
                put("franja", franjaHoraria)
                put("notas", notas)
                put("zona", zona)
                put("estado", "PENDIENTE")
                put("created_at", System.currentTimeMillis())
            }
            newSolicitudId = db.insert(TABLE_SOLICITUDES, null, cvSolicitud)

            if (newSolicitudId == -1L) throw Exception("Error al insertar la solicitud.")

            db.setTransactionSuccessful()
        } catch (e: Exception) {
            Log.e(TAG, "Transacción fallida al crear solicitud: ${e.message}")
            newSolicitudId = -1L
        } finally {
            db.endTransaction()
            db.close()
        }
        return newSolicitudId
    }

    // ==========================================================
    // 2. LECTURA (Listados Enriquecidos para Adaptador)
    // ==========================================================

    // 🏆 CLASE DE DATOS REQUERIDA POR EL ADAPTADOR (Agregada para que SolicitudAdapter funcione)
    data class SolicitudItem(
        val id: Long,
        val estado: String,
        val direccion: String, // Usado para tvAddress
        val fecha: String,    // Usado para tvDate
        val zona: String,
        val franjaHoraria: String,
        val idCliente: Long
    )

    // ... (El resto de tu código de lectura, actualización y mapeo es correcto y se mantiene) ...

    /**
     * Obtiene una lista ENRIQUECIDA de solicitudes por ID de Usuario (Cliente).
     * El resultado se mapea al modelo completo [Solicitud] para el adaptador.
     */
    fun getSolicitudesEnrichedByUserId(userId: Long): List<Solicitud> {
        return executeEnrichedQuery(
            whereClause = "s.user_id = ?",
            whereArgs = arrayOf(userId.toString())
        )
    }

    /**
     * Obtiene una lista ENRIQUECIDA de solicitudes asignadas a un recolector/conductor.
     *
     * @param recolectorId El ID del recolector/conductor.
     * @return Lista de objetos [Solicitud] para el adaptador.
     */
    fun getSolicitudesEnrichedByRecolectorId(recolectorId: Long): List<Solicitud> {
        return executeEnrichedQuery(
            // El conductor solo ve las rutas que no han sido finalizadas o canceladas.
            whereClause = "s.recolector_id = ? AND s.estado NOT IN ('ENTREGADA', 'CANCELADA', 'FINALIZADA')",
            whereArgs = arrayOf(recolectorId.toString())
        )
    }

    /**
     * Obtiene una lista ENRIQUECIDA de solicitudes ASIGNADAS o en ruta, filtradas por ZONA.
     * Este método es usado por el Gestor/Funcionario para ver las órdenes en curso.
     *
     * @param zona La zona para la cual se buscan solicitudes asignadas.
     * @return Lista de objetos [Solicitud] enriquecidos.
     */
    fun getSolicitudesAsignadasEnriquecidasPorZona(zona: String): List<Solicitud> {
        return executeEnrichedQuery(
            whereClause = "s.zona = ? AND s.estado IN ('ASIGNADA', 'EN_RECOLECCION', 'RECOGIDA')",
            whereArgs = arrayOf(zona)
        )
    }

    /**
     * Obtiene una lista ENRIQUECIDA de solicitudes PENDIENTES, filtradas por ZONA.
     *
     * @param zona La zona para la cual se buscan solicitudes pendientes de asignación.
     * @return Lista de objetos [Solicitud] enriquecidos.
     */
    fun getSolicitudesPendientesEnriquecidasPorZona(zona: String): List<Solicitud> {
        return executeEnrichedQuery(
            whereClause = "s.zona = ? AND s.estado = 'PENDIENTE'",
            whereArgs = arrayOf(zona)
        )
    }

    /**
     * Obtiene una lista ENRIQUECIDA de solicitudes FINALIZADAS (ENTREGADA, CANCELADA, FINALIZADA)
     * filtradas por ZONA. Usado para el historial del Gestor/Funcionario.
     *
     * @param zona La zona para la cual se buscan solicitudes completadas.
     * @return Lista de objetos [Solicitud] enriquecidos.
     */
    fun getSolicitudesFinalizadasEnriquecidasPorZona(zona: String): List<Solicitud> {
        return executeEnrichedQuery(
            whereClause = "s.zona = ? AND s.estado IN ('ENTREGADA', 'CANCELADA', 'FINALIZADA')",
            whereArgs = arrayOf(zona)
        )
    }


    /**
     * Implementación centralizada de la consulta JOIN para obtener el modelo [Solicitud] completo.
     */
    private fun executeEnrichedQuery(
        whereClause: String,
        whereArgs: Array<String>
    ): List<Solicitud> {
        val solicitudes = mutableListOf<Solicitud>()
        val db = helper.readableDatabase
        var cursor: Cursor? = null

        val query = """
            SELECT 
                s.id, s.user_id, s.recolector_id, s.direccion_id, s.fecha, s.franja, s.notas, s.zona, s.guia_id, s.estado, s.confirmation_code, s.created_at,
                d.full_address, d.ciudad,
                u.name AS client_name,
                g.tracking_number, g.descripcion, g.valor, g.peso
            FROM 
                $TABLE_SOLICITUDES s
            INNER JOIN 
                $TABLE_DIRECCIONES d ON s.direccion_id = d.id
            INNER JOIN
                $TABLE_USERS u ON s.user_id = u.id
            LEFT JOIN
                $TABLE_GUIA g ON s.guia_id = g.id
            WHERE 
                $whereClause
            ORDER BY 
                s.created_at DESC
        """.trimIndent()

        try {
            cursor = db.rawQuery(query, whereArgs)
            while (cursor.moveToNext()) {
                solicitudes.add(cursorToSolicitudEnriquecida(cursor))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error al ejecutar consulta enriquecida: ${e.message}")
        } finally {
            cursor?.close()
            db.close()
        }
        return solicitudes
    }

    // ==========================================================
    // 3. ACTUALIZACIÓN (Gestor/Conductor)
    // ==========================================================

    /**
     * Actualiza el estado de una solicitud.
     */
    fun actualizarEstado(solicitudId: Long, newState: String, recolectorId: Long? = null): Int {
        val db = helper.writableDatabase
        var rowsAffected = 0

        val cv = ContentValues().apply {
            put("estado", newState)
        }

        // El conductor solo puede actualizar su propia ruta, el gestor/admin puede actualizar cualquiera.
        val whereClause = if (recolectorId != null) {
            "id = ? AND recolector_id = ?"
        } else {
            "id = ?"
        }

        val whereArgs = if (recolectorId != null) {
            arrayOf(solicitudId.toString(), recolectorId.toString())
        } else {
            arrayOf(solicitudId.toString())
        }


        try {
            rowsAffected = db.update(TABLE_SOLICITUDES, cv, whereClause, whereArgs)
        } catch (e: Exception) {
            Log.e(TAG, "Error al actualizar estado: ${e.message}")
        } finally {
            db.close()
        }
        return rowsAffected
    }

    /**
     * Asigna una solicitud pendiente a un recolector/conductor específico.
     */
    fun asignarRecolector(solicitudId: Long, recolectorId: Long): Int {
        val db = helper.writableDatabase
        var rowsAffected = 0

        val cv = ContentValues().apply {
            put("recolector_id", recolectorId)
            put("estado", "ASIGNADA")
        }

        val whereClause = "id = ? AND estado = 'PENDIENTE'"
        val whereArgs = arrayOf(solicitudId.toString())

        try {
            rowsAffected = db.update(TABLE_SOLICITUDES, cv, whereClause, whereArgs)
        } catch (e: Exception) {
            Log.e(TAG, "Error al asignar recolector: ${e.message}")
        } finally {
            db.close()
        }
        return rowsAffected
    }

    // ==========================================================
    // 4. MAPPING (Cursor a Modelo de Dominio)
    // ==========================================================

    /**
     * Mapea un [Cursor] de la consulta ENRIQUECIDA a un objeto [Solicitud] completo.
     */
    private fun cursorToSolicitudEnriquecida(cursor: Cursor): Solicitud {

        // Columnas de la tabla SOLICITUDES
        val id = cursor.getLong(cursor.getColumnIndexOrThrow("id"))
        val userId = cursor.getLong(cursor.getColumnIndexOrThrow("user_id"))
        val recolectorIdIndex = cursor.getColumnIndexOrThrow("recolector_id")
        val recolectorId =
            if (!cursor.isNull(recolectorIdIndex)) cursor.getLong(recolectorIdIndex) else null
        val direccionId = cursor.getLong(cursor.getColumnIndexOrThrow("direccion_id"))
        val fecha = cursor.getString(cursor.getColumnIndexOrThrow("fecha"))
        val franja = cursor.getString(cursor.getColumnIndexOrThrow("franja"))
        val notas = cursor.getString(cursor.getColumnIndexOrThrow("notas"))
        val zona = cursor.getString(cursor.getColumnIndexOrThrow("zona"))
        val guiaIdIndex = cursor.getColumnIndexOrThrow("guia_id")
        val guiaId = if (!cursor.isNull(guiaIdIndex)) cursor.getLong(guiaIdIndex) else null
        val estado = cursor.getString(cursor.getColumnIndexOrThrow("estado"))
        val confirmationCodeIndex = cursor.getColumnIndexOrThrow("confirmation_code")
        val confirmationCode =
            if (!cursor.isNull(confirmationCodeIndex)) cursor.getString(confirmationCodeIndex) else null
        val createdAt = cursor.getLong(cursor.getColumnIndexOrThrow("created_at"))

        // Columnas de la tabla DIRECCIONES
        val fullAddress = cursor.getString(cursor.getColumnIndexOrThrow("full_address"))
        val ciudad = cursor.getString(cursor.getColumnIndexOrThrow("ciudad"))

        // Columnas de la tabla USERS
        val clientNameIndex = cursor.getColumnIndexOrThrow("client_name")
        val clientName =
            if (!cursor.isNull(clientNameIndex)) cursor.getString(clientNameIndex) else "N/D"

        // Columnas de la tabla GUIA (Nullable)
        val trackingNumberIndex = cursor.getColumnIndexOrThrow("tracking_number")
        val trackingNumber =
            if (!cursor.isNull(trackingNumberIndex)) cursor.getString(trackingNumberIndex) else null
        val descripcionIndex = cursor.getColumnIndexOrThrow("descripcion")
        val descripcion =
            if (!cursor.isNull(descripcionIndex)) cursor.getString(descripcionIndex) else null
        val valorIndex = cursor.getColumnIndexOrThrow("valor")
        val valor = if (!cursor.isNull(valorIndex)) cursor.getDouble(valorIndex) else 0.0
        val pesoIndex = cursor.getColumnIndexOrThrow("peso")
        val peso = if (!cursor.isNull(pesoIndex)) cursor.getDouble(pesoIndex) else 0.0

        return Solicitud(
            id = id,
            userId = userId,
            recolectorId = recolectorId,
            direccionId = direccionId,
            fecha = fecha,
            franja = franja,
            notas = notas,
            zona = zona,
            guiaId = guiaId,
            estado = estado,
            confirmationCode = confirmationCode,
            createdAt = createdAt,
            fullAddress = fullAddress,
            ciudad = ciudad,
            clientName = clientName,
            trackingNumber = trackingNumber,
            descripcion = descripcion,
            valor = valor,
            peso = peso
        )
    }

}